DirectProgramming/DPC++FPGA/Tutorials/Features/ac_types/ac_fixed/src/ac_fixed.cpp

```diff
@@ -0,0 +1,193 @@
+      ac_fixed<20, 10, true, AC_RND, AC_SAT> t(out_acc[0]);
+      ac_fixed<20, 10, true, AC_RND, AC_SAT> some_offset(0.5f);
+      out_acc[0] = t + some_offset;
+    });
+  });
+}
+
+void TestCalculateWithFloat(queue &q, const float x, float &ret) {
+  buffer<float, 1> inp_buffer(&x, 1);
+  buffer<float, 1> ret_buffer(&ret, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp_buffer, h, read_only};
+    accessor res{ret_buffer, h, write_only, no_init};
+
+    h.single_task<CalculateWithFloat>([=] {
+      float sin_ret = sinf(x[0]);
+      float cos_ret = cosf(x[0]);
+      res[0] = sqrtf(sin_ret * sin_ret + cos_ret * cos_ret);
+    });
+  });
+}
+
+// clang-format off
+// Please refer to ac_fixed_math.hpp header file for fixed point math
+// functions' type deduction rule. In this case, following those rules:
+// I, W, S are input type template parameter (ac_fixed<W, I, S>)
+// rI, rW, rS are output type template parameter (ac_fixed<rW, rI, rS>)
+//* Function Name         Type Propagation Rule
+//* sqrt_fixed             rI = I, rW = W, rS = S
+//* sin_fixed              For signed (S == true), rI == 2, rW =  W - I + 2;
+//*                        For unsigned (S == false), I == 1, rW =  W - I + 1
+//* cos_fixed              For signed (S == true), rI == 2, rW =  W - I + 2;
+//*                        For unsigned (S == false), I == 1, rW =  W - I + 1
+// clang-format on
+void TestCalculateWithACFixed(queue &q, const fixed_10_3_t &x,
+                              fixed_9_2_t &ret) {
+  buffer<fixed_10_3_t, 1> inp_buffer(&x, 1);
+  buffer<fixed_9_2_t, 1> ret_buffer(&ret, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp_buffer, h, read_only};
+    accessor res{ret_buffer, h, write_only, no_init};
+
+    h.single_task<CalculateWithACFixed>([=] {
+      fixed_9_2_t sin_ret = sin_fixed(x[0]);
+      fixed_9_2_t cos_ret = cos_fixed(x[0]);
+      res[0] = sqrt_fixed(sin_ret * sin_ret + cos_ret * cos_ret);
+    });
+  });
+}
+
+constexpr int SIZE = 5;
+
+int main() {
+#if defined(FPGA_EMULATOR)
+  ext::intel::fpga_emulator_selector selector;
+#else
+  ext::intel::fpga_selector selector;
+#endif
+
+  try {
+    // Create the SYCL device queue
+    queue q(selector, dpc_common::exception_handler);
+
+    ac_fixed<20, 10, true, AC_RND, AC_SAT> a;
+    ac_fixed<20, 10, true, AC_RND, AC_SAT> b = 3.1415f;
+
+    TestConstructFromFloat(q, 3.1415f, a);
+    std::cout << "Constructed from float:\t\t" << a << "\n";
+
+    TestConstructFromACFixed(q, b);
+    std::cout << "Constructed from ac_fixed:\t" << b << "\n\n";
+
+    constexpr float inputs[SIZE] = {-0.807991899423f, -2.09982907558f,
+                                    -0.742066235466f, -2.33217071676f,
+                                    1.14324158042f};
+
+    // quantum: the minimum positive value this type can represent
+    // Quantum is 1 / 2 ^ (W - I), where W and I are the total width and the
+    // integer width of the ac_fixed number
+    constexpr fixed_9_2_t quantum = 0.0078125f;
```

Comment:
I'm surprised there is not a utitlity function for determining this from the `ac_fixed` API